File: frontend/src/types/plot_bridge.ts
/**
 * V4.1 K2 桥段四章结构 - 前端类型定义
 *
 * 对应后端 PlotBridge model（详见 backend/app/models/plot_bridge.py）
 */

type BridgePosition = 'intro' | 'build' | 'payoff' | 'aftermath';

/** 题材族模板 key（对应 backend/app/services/bridge_templates.py） */
type BridgeTemplateKey = 'showoff' | 'mystery' | 'romance' | 'infinite';

export const BRIDGE_TEMPLATE_UI: Record<
  BridgeTemplateKey,
  { name: string; payoffLabel: string; positions: Record<BridgePosition, string>; hints: Record<BridgePosition, string> }
> = {
  showoff: {
    name: '爽文装逼打脸流',
    payoffLabel: '装逼点',
    positions: { intro: 'C1 代入', build: 'C2 拉扯', payoff: 'C3 兑现', aftermath: 'C4 善后' },
    hints: { intro: '5:5', build: '9:1 章尾开装', payoff: '无钩子', aftermath: '承上启下' },
  },
  mystery: {
    name: '悬疑反转流',
    payoffLabel: '反转点',
    positions: { intro: 'C1 异象疑点', build: 'C2 追查误导', payoff: 'C3 反转揭示', aftermath: 'C4 余波新疑' },
    hints: { intro: '5:5', build: '9:1 章尾关键线索', payoff: '无钩子', aftermath: '承上启下' },
  },
  romance: {
    name: '言情推拉流',
    payoffLabel: '情感兑现点',
    positions: { intro: 'C1 情感缺口', build: 'C2 推拉误会', payoff: 'C3 情感兑现', aftermath: 'C4 新状态' },
    hints: { intro: '5:5', build: '9:1 章尾心动动作', payoff: '无钩子', aftermath: '承上启下' },
  },
  infinite: {
    name: '无限流规则破局',
    payoffLabel: '破局点',
    positions: { intro: 'C1 入局规则', build: 'C2 试探代价', payoff: 'C3 破局兑现', aftermath: 'C4 结算' },
    hints: { intro: '5:5', build: '9:1 章尾看破规则', payoff: '无钩子', aftermath: '承上启下' },
  },
};

/** 把后端返回的 template 字符串收敛成已知 key（未知 / 缺失 → showoff） */
export function resolveTemplateKey(key: string | null | undefined): BridgeTemplateKey {
  return key && key in BRIDGE_TEMPLATE_UI ? (key as BridgeTemplateKey) : 'showoff';
}

/** 桥段填充溯源（plot_bridges.generation_meta，后端 build_fill_provenance 产出） */
export interface BridgeGenerationMeta {
  version: number;
  generated_at: string;
  scene: string;
  model: string;
  model_tier: string;
  template: string;
  reference_pack: { title: string; dimensions: Record<string, string> } | null;
  slots: { filled: string[]; truncated: string[]; skipped: string[] };
  inputs: {
    story_outline_fields: string[];
    beat: { index: number; title: string };
    next_beat_title: string | null;
    prev_bridge_number: number | null;
    ledger_bridge_numbers: number[];
    opening_rules: boolean;
    bridge_numbers: number[];
  };
  tokens_estimate: number;
  warnings: string[];
}

/** fill-stream 的 meta 事件（每个子批一次） */
export interface FillMetaEvent {
  type: 'meta';
  beat_index: number;
  bridge_numbers: number[];
  provenance: BridgeGenerationMeta;
  seq?: number;
}

/** 打字机快照：LLM 已写出的字段（半截 JSON 容错解析结果） */
export type PartialBridge = { bridge_number: number } & Partial<
  Pick<
    PlotBridge,
    'title' | 'goal' | 'showoff_point' | 'golden_finger_usage' | 'c1_intro' | 'c2_build' | 'c3_payoff' | 'c4_aftermath' | 'next_bridge_hook'
  >
>;

export interface FillPartialEvent {
  type: 'partial';
  beat_index: number;
  bridge_numbers: number[];
  bridges: PartialBridge[];
  content_chars: number;
  elapsed: number;
  seq?: number;
}

/** 推理模型思考中 / 网络静默心跳：只有计数与计时，不含思考文本 */
export interface FillThinkingEvent {
  type: 'thinking';
  beat_index: number;
  bridge_numbers: number[];
  reasoning_chars: number;
  content_chars: number;
  elapsed: number;
  seq?: number;
}

/** 一个子批落库后的完整桥段（卡片 draft → ready） */
export interface FillBridgesEvent {
  type: 'bridges';
  beat_index: number;
  bridges: PlotBridge[];
  seq?: number;
}

export type BridgeStatus = 'draft' | 'ready' | 'generating' | 'completed';

export const BRIDGE_STATUS_LABEL: Record<BridgeStatus, string> = {
  draft: '草稿',
  ready: '就绪',
  generating: '生成中',
  completed: '已展开',
};

/** 副线任务：支线/角色线节点挂到主线桥段（由后端 bridge_slot_planner 计算） */
export interface SecondaryBeatTask {
  plot_line_id: string;
  line_title: string;
  line_type: string;
  beat_index: number;
  beat_title: string;
  beat_description: string;
  coverage_start: number;
  coverage_end: number;
  /** primary = 本桥段主 B 线（须推进）；mention = 保温提及一句。旧数据缺省按 primary */
  role?: 'primary' | 'mention';
  /** offset = 与主线错峰；merge = 汇入主线节点兑现桥段 */
  relation?: 'offset' | 'merge';
}

type LineMode = 'companion' | 'inserted' | 'converge';

export const LINE_MODE_LABEL: Record<LineMode, string> = {
  companion: '伴生',
  inserted: '插入',
  converge: '汇流',
};

/** 一条副线在本次规划中的预算账（plan-preview.line_budgets） */
export interface LineBudget {
  plot_line_id: string;
  line_title: string;
  line_type: string;
  anchored: boolean;
  mode: LineMode | null;
  anchor_start_beat: number | null;
  anchor_end_beat: number | null;
  estimated_chapters: number | null;
  /** 锚定线：round(预算章数/4)；未锚定线：0（不限） */
  primary_quota: number;
  primary_bridges: number;
  mention_bridges: number;
}

export const isPrimaryTask = (t: SecondaryBeatTask) => (t.role ?? 'primary') !== 'mention';

export interface PlotBridge {
  id: string;
  project_id: string;
  bridge_number: number;
  title: string;
  goal: string;
  showoff_point: string;
  golden_finger_usage: string | null;
  c1_intro: string | null;
  c2_build: string | null;
  c3_payoff: string | null;
  c4_aftermath: string | null;
  next_bridge_hook: string | null;
  status: BridgeStatus;
  order_index: number | null;
  // 桥段 ↔ 主线节点绑定字段（代码写入，非 LLM）
  plot_line_id: string | null;
  beat_index: number | null;
  beat_coverage_start: number | null;
  beat_coverage_end: number | null;
  /** 副线任务列表 */
  secondary_beats: SecondaryBeatTask[];
  /** 确定性章号范围：第 4(n-1)+1 … 4n 章 */
  chapter_start: number;
  chapter_end: number;
  /** 最近一次 LLM 填充的溯源；draft / 旧数据为 null */
  generation_meta?: BridgeGenerationMeta | null;
  /** 填充时使用的题材模板 key；null 时按 showoff 展示 */
  template?: string | null;
}

/** GET /projects/{id}/bridges/plan-preview 返回的槽位表（纯计算，不写库） */
export interface BridgeSlotPreview {
  main_line_id: string;
  total_bridges: number;
  total_chapters: number;
  /** beat_index(字符串) → 该节点桥段数 */
  beat_quotas: Record<string, number>;
  slots: Array<{
    bridge_number: number;
    plot_line_id: string;
    beat_index: number;
    beat_title: string;
    beat_description: string;
    beat_weight: number;
    coverage_start: number;
    coverage_end: number;
    chapter_start: number;
    chapter_end: number;
    secondary: SecondaryBeatTask[];
  }>;
  /** 每条副线的预算账 */
  line_budgets: LineBudget[];
}

/** POST /projects/{id}/bridges/fill-stream 请求体 */
export interface FillBridgesRequest {
  model?: string;
  /** 只填充该主线节点的 draft 桥段 */
  beat_index?: number;
}

/** fill-stream 的 result 事件 */
export interface FillBridgesResult {
  type: 'done';
  filled: number;
  remaining_drafts: number;
}

export interface ExpandBridgeRequest {
  model?: string;
}

export interface ExpandBridgeResponse {
  success: boolean;
  bridge_id: string;
  chapter_count: number;
  chapter_ids: string[];
}

export interface UpdateBridgeRequest {
  title?: string;
  goal?: string;
  showoff_point?: string;
  golden_finger_usage?: string;
  c1_intro?: string;
  c2_build?: string;
  c3_payoff?: string;
  c4_aftermath?: string;
  next_bridge_hook?: string;
  status?: BridgeStatus;
}

/** 批量展开请求（按 bridge_number 顺序，首个失败即停止）。 */
export interface ExpandAllBridgesRequest {
  model?: string;
}

/** 批量展开结果（后台任务 result 事件的 data；服务层汇总字典，没有 success 字段）。 */
export interface ExpandAllBridgesResponse {
  success?: boolean;
  total: number;
  succeeded: string[];
  failed: Array<{ bridge_id: string; error: string }>;
  created_chapter_count: number;
}

File: backend/app/models/plot_bridge.py
"""V4.1 K2 桥段四章结构 - PlotBridge 模型（详见 v4_design.md §3.2.1）。

设计：一个桥段 ≈ 4 章，每章在桥段内有明确位置（C1/C2/C3/C4）。

字段：
- 桥段意图：goal / showoff_point / golden_finger_usage
- 4 章内容卡：c1_intro / c2_build / c3_payoff / c4_aftermath（4 段文本提示，
  桥段规划阶段 AI 生成，章纲展开阶段使用）
- 上下文衔接：prev_bridge_id / next_bridge_hook
- 状态：draft / ready / generating / completed

关联：
- 一个项目下 N 个桥段，按 order_index 排序
- 一个桥段挂 4 个 chapter_outline（chapter_outlines.bridge_id 反向关联）
- 可选属于一条 plot_line
"""
import uuid

from sqlalchemy import Column, DateTime, Float, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db_base import Base


class PlotBridge(Base):
    """K2 桥段表 - 一个桥段约等于 4 章。"""

    __tablename__ = "plot_bridges"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(
        String(36),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    plot_line_id = Column(
        String(36),
        ForeignKey("plot_lines.id", ondelete="SET NULL"),
        nullable=True,
        comment="所属剧情线（可选，桥段可独立存在）",
    )

    # ---- V4.1 K2 分层契合：桥段 ↔ 剧情线节点（beat）绑定 ----
    # 设计：方案 C（分层契合）— 桥段是「主线节点的展开形式」，
    # 一个 beat 按 weight × estimated_chapters 分配章数，再按 4 章/桥段切成 N 个桥段。
    # 这些字段在 by_plot_line 规划模式下由 LLM 填充；free 模式下保持 NULL（向后兼容）。
    beat_index = Column(
        Integer,
        nullable=True,
        comment="所属节点 index（对应 PlotLine.timeline_data.beats[].index）；NULL 表示桥段不绑节点",
    )
    beat_coverage_start = Column(
        Float,
        nullable=True,
        comment="本桥段覆盖该节点的起始进度（0.0-1.0），用于章纲生成时推进 beat coverage",
    )
    beat_coverage_end = Column(
        Float,
        nullable=True,
        comment="本桥段覆盖该节点的结束进度（0.0-1.0）",
    )
    secondary_beats = Column(
        Text,
        nullable=True,
        comment="副线任务 JSON 数组：[{plot_line_id, line_title, line_type, beat_index, beat_title, beat_description, coverage_start, coverage_end}]，由 bridge_slot_planner 计算写入",
    )

    bridge_number = Column(
        Integer,
        nullable=False,
        comment="桥段序号（项目内全局递增）",
    )
    title = Column(
        String(200),
        nullable=False,
        comment="桥段标题，如『拜师云鹿书院』",
    )

    # ---- 核心字段：桥段意图 ----
    goal = Column(
        Text,
        nullable=False,
        comment="本桥段要解决的具体问题，如『求大儒收留家人』",
    )
    showoff_point = Column(
        Text,
        nullable=False,
        comment="装逼/爽点设计，如『即兴一首劝学诗征服大儒』",
    )
    payoff_type = Column(
        String(40),
        nullable=True,
        comment="兑现方式枚举（题材模板 payoff_types 之一，如 打脸 / 扮猪吃虎）；账本按类型统计避免长线套路重复",
    )
    golden_finger_usage = Column(
        Text,
        nullable=True,
        comment="本桥段如何使用金手指（如：诗词储备）",
    )

    # ---- 4 章内容卡（JSON 存储，简化挂载）----
    c1_intro = Column(
        Text,
        nullable=True,
        comment="C1 代入+信息差 设计：上半日常代入素材、下半信息差展示",
    )
    c2_build = Column(
        Text,
        nullable=True,
        comment="C2 拉扯+开装 设计：拉扯素材、章尾开装的具体动作",
    )
    c3_payoff = Column(
        Text,
        nullable=True,
        comment="C3 兑现爽点 设计：装逼的完整展开、配角反应",
    )
    c4_aftermath = Column(
        Text,
        nullable=True,
        comment="C4 善后+下一目标 设计：本桥段收尾事件、下一桥段引子",
    )

    # ---- 上下文衔接 ----
    prev_bridge_id = Column(
        String(36),
        ForeignKey("plot_bridges.id"),
        nullable=True,
        comment="上一桥段（用于 C1 代入处理）",
    )
    next_bridge_hook = Column(
        Text,
        nullable=True,
        comment="给下一桥段的钩子（C4 必须写）",
    )

    # ---- 生成溯源（2026-09 评审问题 H）----
    generation_meta = Column(
        Text,
        nullable=True,
        comment="本桥段最近一次 LLM 填充的 provenance JSON：模型/档位/题材模板/参考包维度/槽位填充与截断/业务输入/警告",
    )

    # ---- 状态 ----
    status = Column(
        String(20),
        default="draft",
        comment="draft / ready / generating / completed",
    )
    order_index = Column(Integer, nullable=True)

    # ---- 时间戳 ----
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
    )

    # ---- 关联关系 ----
    chapter_outlines = relationship(
        "ChapterOutline",
        back_populates="bridge",
        foreign_keys="[ChapterOutline.bridge_id]",
    )

    __table_args__ = (
        Index("idx_plot_bridge_project_order", "project_id", "order_index"),
        Index("idx_plot_bridge_project_number", "project_id", "bridge_number"),
    )

    def __repr__(self) -> str:
        return (
            f"<PlotBridge(id={self.id[:8] if self.id else None}..., "
            f"number={self.bridge_number}, title={self.title!r})>"
        )
